test/assignment-5/CollateralizedVault.t.sol

```diff
@@ -0,0 +1,270 @@
+        vault.withdraw(1);
+    }
+
+    function testCanWithdrawEntireCollateralIfPaidAllDebt() public {
+        assertEq(weth.balanceOf(USER), 7 ether);
+        assertEq(vault.borrows(USER), 6000 * 1e18);
+        assertEq(vault.deposits(USER), 3 ether);
+
+        // Repay entire debt
+        vm.startPrank(USER);
+        vault.repay(3 * 2000 ether);
+        // No more debt
+        assertEq(vault.borrows(USER), 0);
+        // But also no more DAI
+        assertEq(dai.balanceOf(USER), 0);
+
+        vault.withdraw(3 ether);
+        // WETH is returned to user
+        assertEq(weth.balanceOf(USER), 10 ether);
+    }
+
+    function testOnlyOwnerCanLiquidate() public {
+        vm.prank(address(0x1234));
+        vm.expectRevert("Ownable: caller is not the owner");
+        vault.liquidate(USER);
+    }
+
+    function testOwnerCantLiquidateIfDebtIsCollateralized() public {
+        vm.expectRevert(CollateralizedVault.UserDebtIsSufficientlyCollateralized.selector);
+        vault.liquidate(USER);
+    }
+
+    function testOwnerCanLiquidateIfDebtIsUnderCollateralized() public {
+        // WETH went down, now only $1000, DAI/ETH = 1/1000
+        priceFeedMock.setPrice(1e18 / 1000);
+
+        // Need 6 WETH
+        assertEq(vault.getRequiredCollateral(vault.borrows(USER)), 6 ether);
+        // But only 3 is deposited
+        assertEq(vault.deposits(USER), 3 ether);
+
+        // liquidate user
+        vault.liquidate(USER);
+
+        assertEq(vault.deposits(USER), 0);
+        assertEq(vault.borrows(USER), 0);
+    }
+}
+
+abstract contract PartiallyRepaidDebtState is BorrowedState {
+    function setUp() virtual override public {
+        super.setUp();
+
+        // Repay 2 thirds of debt
+        vm.prank(USER);
+        vault.repay(2 * 2000 ether);
+    }
+}
+
+contract PartiallyRepaidDebtStateTest is PartiallyRepaidDebtState {
+
+    function testWithdrawPartialCollateral() public {
+        // 1 third debt left
+        assertEq(vault.borrows(USER), 2000 ether);
+
+        vm.prank(USER);
+        vault.withdraw(2 ether);
+        // WETH is returned to user
+        assertEq(weth.balanceOf(USER), 9 ether);
+    }
+
+    function testRevertWhenTryingToWithdrawTooMuch() public {
+        vm.prank(USER);
+        vm.expectRevert(CollateralizedVault.TooMuchDebt.selector);
+        vault.withdraw(2 ether + 1);
+    }
+
+    function testCanWithdrawLessIfPriceMovedNegatively() public {
+        // WETH went down, now only $1000, price = 1/1000
+        priceFeedMock.setPrice(1e18 / 1000);
+
+        // can't withdraw 2 WETH
+        vm.prank(USER);
+        vm.expectRevert(CollateralizedVault.TooMuchDebt.selector);
+        vault.withdraw(2 ether);
+
+        // but 1 WETH is OK
+        vm.prank(USER);
+        vault.withdraw(1 ether);
+    }
+
+    function testCanWithdrawMoreIfPriceMovePositively() public {
+        // WETH went up 25%, now at $2500, price = 1/2500
```

Comment:
generally speaking, these tests are great.  well thought out, ur hitting all the important state transitions.  well organized.

v well done.